external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -360,98 +362,165 @@ pub enum InclusionCheck {
-            // below, the one below should be kept if we allow adding variants in subset
-            // mode.
-            if old_enum.variants.len() != new_enum.variants.len() {
-                return err;
+        }
+        // enum checks
+        compare_identifiers(old_module.enums.keys(), new_module.enums.keys()).for_each(|mark| {
+            match mark {
+                Mark::New(name) => context.enum_new(name, new_module.enums.get(name).unwrap()),
+                Mark::Missing(name) => context.enum_missing(name),
+                Mark::Existing(name) => {
+                    let old_enum = old_module.enums.get(name).unwrap();
+                    let new_enum = new_module.enums.get(name).unwrap();
+                    if old_enum != new_enum {
+                        context.enum_change(name, old_enum);
+                    }
+                }
             }
+        });
 
-            if self == &Self::Equal && old_enum.variants.len() != new_enum.variants.len() {
-                return err;
-            }
-            // NB: We are using the fact that the variants are sorted by tag, that we've
-            // already ensured that the old variants are >= new variants, and the fact
-            // that zip will truncate the second iterator if there are extra there to allow
-            // adding new variants to enums in `Self::Subset` compatibility mode.
-            if !old_enum
-                .variants
-                .iter()
-                .zip(&new_enum.variants)
-                .all(|(old, new)| old == new)
-            {
-                return err;
-            }
-        }
+        // function checks
 
-        // Function checks
-        for (name, old_func) in &old_module.functions {
-            match new_module
-                .functions
-                .get(name)
-                .or_else(|| new_module.functions.get(name))
-            {
-                Some(new_func) if old_func == new_func => (),
-                _ => {
-                    return err;
+        compare_identifiers(old_module.functions.keys(), new_module.functions.keys()).for_each(
+            |mark| match mark {
+                Mark::New(name) => {
+                    context.function_new(name, new_module.functions.get(name).unwrap())
+                }
+                Mark::Missing(name) => context.function_missing(name),
+                Mark::Existing(name) => {
+                    let old_function = old_module.functions.get(name).unwrap();
+                    let new_function = new_module.functions.get(name).unwrap();
+                    if old_function != new_function {
+                        context.function_change(name, old_function);
+                    }
+                }
+            },
+        );
+
+        // use iters for friends
+        let mut old_friend_iter = old_module.friends.iter().peekable();
+        let mut new_friend_iter = new_module.friends.iter().peekable();
+        loop {
```

Comment:
I kind of want to suggest generalising `compare_identifier` so that it can be used here -- it should be fairly simple to do, -- it just needs to accept a comparator function to do the ordering with.